src/status.rs

```diff
@@ -1,166 +1,156 @@
-use std::fmt;
-
+use bytes::Bytes;
 use h2;
-use http::header::HeaderValue;
+use http::HeaderMap;
+use http::header::{HeaderValue, InvalidHeaderValueBytes};
 
 #[derive(Debug, Clone)]
 pub struct Status {
     code: Code,
+    error_message: Option<Bytes>,
+    binary_error_details: Option<Bytes>,
 }
 
-#[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Code(Code_);
-
 impl Status {
-    #[inline]
-    pub fn code(&self) -> Code {
-        self.code
+    pub fn with_code(code: Code) -> Status {
+        Status {
+            code,
+            error_message: None,
+            binary_error_details: None,
+        }
     }
 
-    pub const OK: Status = Status {
-        code: Code(Code_::Ok),
-    };
-
-    pub const CANCELED: Status = Status {
-        code: Code(Code_::Canceled),
-    };
-
-    pub const UNKNOWN: Status = Status {
-        code: Code(Code_::Unknown),
-    };
-
-    pub const INVALID_ARGUMENT: Status = Status {
-        code: Code(Code_::InvalidArgument),
-    };
-
-    pub const DEADLINE_EXCEEDED: Status = Status {
-        code: Code(Code_::DeadlineExceeded),
-    };
-
-    pub const NOT_FOUND: Status = Status {
-        code: Code(Code_::NotFound),
-    };
-
-    pub const ALREADY_EXISTS: Status = Status {
-        code: Code(Code_::AlreadyExists),
-    };
-
-    pub const PERMISSION_DENIED: Status = Status {
-        code: Code(Code_::PermissionDenied),
-    };
-
-    pub const RESOURCE_EXHAUSTED: Status = Status {
-        code: Code(Code_::ResourceExhausted),
-    };
-
-    pub const FAILED_PRECONDITION: Status = Status {
-        code: Code(Code_::FailedPrecondition),
-    };
+    pub fn from_header_map(header_map: &mut HeaderMap) -> Option<Status> {
+        if let Some(code) = header_map.remove("grpc-status") {
+            let code = ::Code::from_bytes(code.as_ref());
+            let error_message = header_map.remove("grpc-message").map(|h| h.into());
+            let binary_error_details = header_map.remove("grpc-status-details-bin").map(|h| h.into());
```

Comment:
Another nit: the string literals `"grpc-message"`, `"grpc-status"`, and `"grpc-status-details-bin"` show up several times in this PR. It would be nice if these could be constants...

Reply:
Done